cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -597,19 +602,149 @@ public void getStatuses(HttpRequest request, HttpResponder responder,
+   * parameters have a valid status will have the status field but all JsonObjects for which the parameters do not have
+   * a valid status will have an error message and statusCode.
+   * </p><p>
+   * For example, if there is no App2 in the data above, then the response would be 200 OK with following possible data:
+   * </p>
+   * <pre><code>
+   * [{"appId": "App1", "programType": "Service", "programId": "Service1", "statusCode": 200},
+   * {"appId": "App1", "programType": "Mapreduce", "programId": "Mapreduce2", "statusCode": 200},
+   * {"appId":"App2", "programType":"Flow", "programId":"Flow1", "statusCode":404, "error": "App: App2 not found"}]
+   * </code></pre>
+   */
+  @POST
+  @Path("/stop")
+  public void stopPrograms(HttpRequest request, HttpResponder responder,
+                           @PathParam("namespace-id") String namespaceId) throws BadRequestException, IOException {
+
+    List<BatchProgram> programs = validateAndGetBatchInput(request, BATCH_PROGRAMS_TYPE);
+
+    List<ListenableFuture<BatchProgramResult>> issuedStops = new ArrayList<>(programs.size());
+    for (final BatchProgram program : programs) {
+      Id.Program programId = Id.Program.from(namespaceId, program.getAppId(),
+                                             program.getProgramType(), program.getProgramId());
+      try {
+        ListenableFuture<BatchProgramResult> issuedStop = Futures.transform(issueStop(programId, null),
+          new Function<ProgramController, BatchProgramResult>() {
+            @Override
+            public BatchProgramResult apply(ProgramController input) {
+              return new BatchProgramResult(
+                program, AppFabricServiceStatus.OK.getCode().getCode(), AppFabricServiceStatus.OK.getMessage());
+            }
+          });
+        issuedStops.add(issuedStop);
+      } catch (NotFoundException e) {
+        issuedStops.add(Futures.immediateFuture(
+          new BatchProgramResult(program, HttpResponseStatus.NOT_FOUND.getCode(), e.getMessage())));
+      } catch (BadRequestException e) {
+        issuedStops.add(Futures.immediateFuture(
+          new BatchProgramResult(program, HttpResponseStatus.BAD_REQUEST.getCode(), e.getMessage())));
+      }
+    }
+
+    List<BatchProgramResult> output = new ArrayList<>(programs.size());
+    // need to keep this index in case there is an exception getting the future, since we won't have the program
+    // information in that scenario
+    int i = 0;
+    for (ListenableFuture<BatchProgramResult> issuedStop : issuedStops) {
+      try {
+        output.add(issuedStop.get());
+      } catch (Throwable t) {
+        LOG.warn(t.getMessage(), t);
+        output.add(new BatchProgramResult(programs.get(i),
+                                          AppFabricServiceStatus.INTERNAL_ERROR.getCode().getCode(),
+                                          AppFabricServiceStatus.INTERNAL_ERROR.getMessage()));
```

Comment:
for the last parameter to BatchProgramResult constructor, shouldn't we pass in t.getMessage(), instead of INTERNAL_ERROR.getMessage())?